util/syspolicy/setting/snapshot.go

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package setting
+
+import (
+	"slices"
+	"strings"
+
+	xmaps "golang.org/x/exp/maps"
+	"tailscale.com/util/deephash"
+)
+
+// Snapshot is an immutable collection of ([Key], [RawItem]) pairs, representing
+// a set of policy settings applied at a specific moment in time.
+// A nil pointer to [Snapshot] is valid.
+type Snapshot struct {
+	m       map[Key]RawItem
+	sig     deephash.Sum // of m
+	summary Summary
+}
+
+// NewSnapshot returns a new [Snapshot] with the specified items and options.
+func NewSnapshot(items map[Key]RawItem, opts ...SummaryOption) *Snapshot {
+	return &Snapshot{m: items, sig: deephash.Hash(&items), summary: SummaryWith(opts...)}
+}
+
+type keyItemPair struct {
+	Key  Key
+	Item RawItem
+}
+
+// All returns an iterator over [[Key], [RawItem]] key-value pairs in b. The
+// iteration order is not specified and is not guaranteed to be the same from
+// one call to the next.
+func (s *Snapshot) All() []keyItemPair {
+	if s == nil {
+		return nil
+	}
+	// TODO(nickkhyl): return iter.Seq2[[Key], [RawItem]] in Go 1.23,
+	// and remove [keyItemPair].
+	items := make([]keyItemPair, 0, len(s.m))
+	for k, i := range s.m {
+		items = append(items, keyItemPair{k, i})
+	}
+	return items
+}
+
+// Get returns the value of the policy setting with the specified key
+// or nil if it does not exist or could not be read.
```

Comment:
or nil upon failure?

Reply:
I don't consider an unconfigured policy setting to be a failure. In fact, most policy settings in most environments should remain unconfigured. That's why I intentionally differentiate between "does not exist" and "could not be read".

But I agree that the current comment is awkward. How about "is not configured or has an error"?